enhancements/sig-architecture/32-extensiblescheduling/32-extensiblescheduling.md

```diff
@@ -0,0 +1,426 @@
+
+- First schedule
+
+The prioritizer "ResourceRatioCPU" will get customized score "cpuratio" from "default" `AddOnPlacementScore`, and calculate a final score for each managed clusters.
+
+- Reschedule
+
+The scores inside `AddOnPlacementScore` are supposed to be updated frequently. Rather than watch the changes in `AddOnPlacementScore` and trigger reschedule, the scheduler framework would periodically (eg. every 5 minutes) reschedule the placement.
+
+**What if no valid score is generated?**
+
+What if some of the scores are invalid? For example, the AddOnPlacementScore CR is not created for some managed clusters, the score is missing in some CR, the score is expired. To handle the invalid cases, we add a PreScore() in the Prioritizer interface, to do some pre-check before Score().
+```golang
+// Prioritizer defines a prioritizer plugin that scores each cluster. The score is normalized
+// as floating between 0 and 1.
+type Prioritizer interface {
+	Plugin
+
+	// PreScore() do some preparation work before Score().
+	PreScore(ctx context.Context, placement *clusterapiv1alpha1.Placement, clusters []*clusterapiv1.ManagedCluster) error
+
+	// Score gives the score to a list of the clusters, it returns a map with the key as
+	// the cluster name.
+	Score(ctx context.Context, placement *clusterapiv1alpha1.Placement, clusters []*clusterapiv1.ManagedCluster) (map[string]int64, error)
+}
+```
+
+With the PreSocre() added, the "AddOnScore" type prioritizer scheduling process is as below:
+
+- First schedule
+  - In PreScore(), the prioritizer checks if 80% of the CRs have a valid score. (80% is a reasonable threshold hardcode so far, might make it configurable in the future). 
+    - If yes, the prioritizer will continue to score the clusters. 
+    - If no, terminate this scheduling and wait for the next reschedule.
+  - In Score(), the prioritizer will get customized score from `AddOnPlacementScore`, and calculate a final score for each managed cluster.
+
+- Reschedule
+  - In PreScore(), the prioritizer checks if 80% of the CRs have a valid score.
+    - If yes, the prioritizer will continue to score the clusters. 
+    - If no, remove the managed cluster which has invalid score from the placement decision.
+  - In Score(), the prioritizer will get customized score from `AddOnPlacementScore`, and calculate a final score for each managed cluster.
+
+Future work：there might be a cluster-scoped CRD to configure the threshold of each prioritizer. For example, the frequency to trigger a re-schedule, the threshold to treat the cluster's score as valid.
+
+### How to maintain the lifecycle of the AddOnPlacementScore CRs? 
+Even this is out of scope in this proposal, 2 options to maintain the resources:
```

Comment:
1 . where the 3rd partcy controller resides to maintain the score (both spoke and hub)
2. how to calculate the score. There are multiple options. For instance if it is a controller on hub to maintain the score, it can normalize based on max/min value it collects. We would consider a common library to make it more easy. 
3. When the score should be updated. ValidUntil must be consider for the 3rd party controller
4. When the score should be created? It can be create with the existence of a ManagedCluster, or on demand for the purpose of reducing objects on hub.

Reply:
updated.